pfio/v2/gcs.py

```diff
@@ -0,0 +1,554 @@
+                        obj._CHUNK_SIZE = bs
+
+            elif 'w' in mode:
+                # Create intermediate directories as simulated ones
+                ps = path.split('/')
+                xdi = -1  # xdi means `reverse order index`
+                while len(ps[:xdi]) != 0:
+                    self.__make_simulated_dir('/'.join(ps[:xdi]))
+                    xdi -= 1
+                if 'b' in mode:
+                    obj = BlobWriter(blob, chunk_size=1024 * 1024,
+                                     ignore_flush=self.ignore_flush)
+                else:
+                    obj = io.TextIOWrapper(
+                        BlobWriter(blob, chunk_size=1024 * 1024,
+                                   ignore_flush=True))
+            else:
+                raise RuntimeError(f'Unknown option: {mode}')
+
+            if self.trace:
+                return GCSProfileIOWrapper(obj)
+            else:
+                return obj
+
+    def list(self, prefix: Optional[str] = "", recursive=False, detail=False):
+        """List all objects (and prefixes)
+
+        Although there is not concept of directory in GCS API,
+        common prefixes shows up like directories.
+        """
+        for e in record_iterable("pfio.v2.GoogleCloudStorage:list",
+                                 self._list(prefix, recursive, detail),
+                                 trace=self.trace):
+            yield e
+
+    def _list(self, prefix: Optional[str] = "", recursive=False, detail=False):
+        path = os.path.join(self.cwd, "" if prefix is None else prefix)
+        path = _normalize_key(path)
+        if path == '.':
+            path = ''
+        elif path != '' and not path.endswith('/'):
+            path += '/'
+        if '/../' in path or path.startswith('..'):
+            raise ValueError('Invalid GCS key: {} as {}'.format(prefix, path))
+
+        blobs = self.bucket.list_blobs(prefix=path,
+                                       delimiter=('' if recursive else '/'),
+                                       timeout=self.connect_time)
+        # objects
+        for blob in blobs:
+            if blob.name == path:
+                continue
+
+            if detail:
+                yield ObjectStat(blob, _format_path(blob.name, path))
+            else:
+                yield _format_path(blob.name, path)
+        # folders
+        for prefix in blobs.prefixes:
+            if detail:
+                yield PrefixStat(_format_path(prefix, path))
+            else:
+                yield _format_path(prefix, path)
+
+    def stat(self, path):
+        """Imitate FileStat with S3 Object metadata
+
+        """
+        with record("pfio.v2.GoogleCloudStorage:stat", trace=self.trace):
+            self._checkfork()
+            path = _normalize_key(os.path.join(self.cwd, path))
+
+            blob = self.bucket.get_blob(path)
+            return ObjectStat(blob, self._get_relative_path(blob.name))
```

Comment:
修正漏れを見つけたので、残しておきます。
```suggestion
            return ObjectStat(blob, _format_path(blob.name, path))
```